Cloud Data Provider

kchadha opened the pull request:
### Proposed Changes

Adds a cloud data provider which creates and manages a web socket connection to the cloud data host provided by `www`. The connection is created upon successfully loading a project. 

This is a work in progress towards adding cloud variables to Scratch 3.0. 

This PR does not include:
 - re-attempting a connection upon error
- logic around whether or not the current state indicates that a cloud data connection should be opened (e.g. user info, project info, etc.)
- remix logic
- ability to create new cloud variables in a 3.0 project

#### Related PRs

- LLK/scratch-vm#1703
- LLK/scratch-www#2245

### Reason for Changes

Cloud variables!

### Test Coverage

Manually tested sending and receiving updates to a 2.0 project containing cloud variables when viewing that project from the 3.0 project page.

### Browser Coverage
Check the OS/browser combinations tested (At least 2)

Mac
 * [x] Chrome 
 * [x] Firefox 
 * [x] Safari
 
Windows
 * [ ] Chrome 
 * [ ] Firefox 
 * [ ] Edge
 
Chromebook
 * [ ] Chrome
 
iPad
* [ ] Safari

Android Tablet
* [ ] Chrome

towerofnix commented on src/lib/cloud-provider.js:
```diff
@@ -0,0 +1,124 @@
+        const msg = {};
+        msg.method = methodName;
+        msg.user = this.username;
+        msg.project_id = this.projectId;
```
Should this require some form of authentication? (I.e. like `?x-token`.) The absolute last thing anyone wants is for a user to impersonate some other user in the cloud data log by overriding `msg.user` to be something other than their own username.

joker314 replied:
@towerofnix This looks the same as 2.0 cloud protocol; where one uses cookies (in the upgrade request, I think) to confirm authenticity.

kchadha replied:
@towerofnix, the messages implicitly use session data, and authentication on the server fails without it.

rschamp commented on src/lib/cloud-provider.js:
```diff
@@ -0,0 +1,124 @@
+
+
+class CloudProvider {
+    constructor (cloudHost, vm, user, projectId) {
```
Docs would be good here. Also I would suggest `username` rather than `user` since it could be confusing going between www — in www `user` might be an object.

rschamp commented on src/lib/cloud-provider.js:
```diff
@@ -0,0 +1,124 @@
+                const varData = {
+                    projectId: this.projectId,
+                    varUpdate: {
+                        projectId: message.project_id,
```
A comment here to explain the two different `projectId`'s would be nice.

kchadha replied:
Oops! That was a remnant of old code that I decided wasn't necessary anymore.

rschamp commented on src/lib/vm-manager-hoc.jsx:
```diff
@@ -54,12 +67,14 @@ const vmManagerHOC = function (WrappedComponent) {
                 onLoadedProject: onLoadedProjectProp,
                 projectData,
                 projectId,
+                userName,
```
I think almost everywhere else (including in the new CloudProvider class!) we use `username` without caps — can we stick with that for consistency?

kchadha commented:
@rschamp I added docs and updated the naming/capitalization of 'username'. I also took out the unnecessary/duplicate projectId in the message being posted to the VM.
